File: sdks/python/src/zeroclaw_plugin_sdk/context.py
"""Context module — wraps session, user_identity, and agent_config host functions.

Each function calls the corresponding ``context_*`` host function via Extism
shared memory, deserializes the JSON response, and returns a typed dataclass.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Dict, List

import extism_pdk as pdk


@dataclass
class SessionContext:
    """Runtime session information provided by the host."""

    channel_name: str
    conversation_id: str
    timestamp: str


@dataclass
class UserIdentity:
    """Identity of the user interacting with the agent."""

    username: str
    display_name: str
    channel_user_id: str


@dataclass
class AgentConfig:
    """Agent configuration provided by the host."""

    name: str
    personality_traits: List[str] = field(default_factory=list)
    identity: Dict[str, str] = field(default_factory=dict)


def session() -> SessionContext:
    """Return the current session context from the host.

    Raises ``RuntimeError`` if the host reports an error.
    """
    raw = pdk.host_fn("context_session", "null")
    response = json.loads(raw)
    if isinstance(response, dict) and response.get("error"):
        raise RuntimeError(response["error"])
    return SessionContext(
        channel_name=response["channel_name"],
        conversation_id=response["conversation_id"],
        timestamp=response["timestamp"],
    )


def user_identity() -> UserIdentity:
    """Return the current user identity from the host.

    Raises ``RuntimeError`` if the host reports an error.
    """
    raw = pdk.host_fn("context_user_identity", "null")
    response = json.loads(raw)
    if isinstance(response, dict) and response.get("error"):
        raise RuntimeError(response["error"])
    return UserIdentity(
        username=response["username"],
        display_name=response["display_name"],
        channel_user_id=response["channel_user_id"],
    )


def agent_config() -> AgentConfig:
    """Return the agent configuration from the host.

    Raises ``RuntimeError`` if the host reports an error.
    """
    raw = pdk.host_fn("context_agent_config", "null")
    response = json.loads(raw)
    if isinstance(response, dict) and response.get("error"):
        raise RuntimeError(response["error"])
    return AgentConfig(
        name=response["name"],
        personality_traits=response.get("personality_traits", []),
        identity=response.get("identity", {}),
    )

File: crates/zeroclaw-plugin-sdk/src/context.rs
//! Context module — wraps session, user_identity, and agent_config host functions.
//!
//! Each function calls the corresponding `context_*` host function via Extism
//! shared memory and returns a typed response struct.

use extism_pdk::*;
use serde::Deserialize;
use std::collections::HashMap;

// ---------------------------------------------------------------------------
// Response types (mirror the host-side structs)
// ---------------------------------------------------------------------------

/// Snapshot of the current session context.
#[derive(Debug, Clone, Deserialize)]
pub struct SessionContext {
    /// The channel name the current request originated from (e.g. "telegram", "slack").
    pub channel_name: String,
    /// An opaque conversation/session identifier.
    pub conversation_id: String,
    /// ISO-8601 timestamp of the current request.
    pub timestamp: String,
}

/// Information about the user who triggered the current invocation.
#[derive(Debug, Clone, Deserialize)]
pub struct UserIdentity {
    /// The user's username (e.g. "jdoe").
    pub username: String,
    /// The user's display name (e.g. "Jane Doe").
    pub display_name: String,
    /// A channel-specific identifier for the user.
    pub channel_user_id: String,
}

/// Agent personality and identity configuration.
#[derive(Debug, Clone, Deserialize)]
pub struct AgentConfig {
    /// The agent's configured display name.
    pub name: String,
    /// Personality traits (e.g. "friendly", "concise", "technical").
    pub personality_traits: Vec<String>,
    /// Arbitrary identity fields from the agent configuration.
    pub identity: HashMap<String, String>,
}

// ---------------------------------------------------------------------------
// Host function imports
// ---------------------------------------------------------------------------

#[host_fn]
extern "ExtismHost" {
    fn context_session(input: Json<()>) -> Json<SessionContext>;
    fn context_user_identity(input: Json<()>) -> Json<UserIdentity>;
    fn context_agent_config(input: Json<()>) -> Json<AgentConfig>;
}

// ---------------------------------------------------------------------------
// Public wrapper API
// ---------------------------------------------------------------------------

/// Get the current session context (channel, conversation ID, timestamp).
pub fn session() -> Result<SessionContext, Error> {
    let Json(ctx) = unsafe { context_session(Json(()))? };
    Ok(ctx)
}

/// Get the identity of the user who triggered this invocation.
pub fn user_identity() -> Result<UserIdentity, Error> {
    let Json(identity) = unsafe { context_user_identity(Json(()))? };
    Ok(identity)
}

/// Get the agent's personality and identity configuration.
pub fn agent_config() -> Result<AgentConfig, Error> {
    let Json(config) = unsafe { context_agent_config(Json(()))? };
    Ok(config)
}
